libs/designer/src/lib/ui/mcp/panel/operation/EditOperationPanelInner.tsx

```diff
@@ -17,62 +26,173 @@ export const EditOperationPanelInner = () => {
   const styles = useMcpPanelStyles();
   const [isDirty, setIsDirty] = useState(false);
 
+  const editOperationRef = useRef<EditOperationRef>(null);
+
+  const handleDismiss = useCallback(() => {
+    dispatch(closePanel());
+  }, [dispatch]);
+
+  const { selectedOperationId, inputParameters } = useSelector((state: RootState) => ({
+    selectedOperationId: state.connector.selectedOperationId,
+    inputParameters: state.operation.inputParameters,
+  }));
+
+  const parameters = selectedOperationId ? inputParameters[selectedOperationId] : null;
+
   const INTL_TEXT = {
     title: intl.formatMessage({
-      id: 'unVPy/',
-      defaultMessage: 'Edit Operations',
-      description: 'Title for edit operations panel',
+      id: 'KDsdC6',
+      defaultMessage: 'Edit Operation',
+      description: 'Title for edit operation panel',
     }),
     closeAriaLabel: intl.formatMessage({
       id: 'kdCuJZ',
       defaultMessage: 'Close panel',
       description: 'Aria label for close button',
     }),
+    unsavedChangesMessage: intl.formatMessage({
+      id: '2uxQXJ',
+      defaultMessage: 'You have unsaved changes. Are you sure you want to discard them?',
+      description: 'Confirmation message for discarding unsaved changes',
+    }),
   };
 
-  const handleDismiss = useCallback(() => {
-    dispatch(closePanel());
-  }, [dispatch]);
+  const resetToOriginalState = useCallback(() => {
+    editOperationRef.current?.resetLocalChanges();
+    setIsDirty(false);
+  }, []);
 
-  const updateOperationData = useCallback(
-    (_operationData: any) => {
-      setIsDirty(true);
-      //TODO: update operation data logic
-    },
-    [setIsDirty]
-  );
+  const handleCancel = useCallback(() => {
+    if (isDirty) {
+      const shouldDiscard = window.confirm(INTL_TEXT.unsavedChangesMessage);
+      if (!shouldDiscard) {
+        return;
+      }
+    }
+
+    resetToOriginalState();
+    handleDismiss();
+  }, [isDirty, INTL_TEXT.unsavedChangesMessage, resetToOriginalState, handleDismiss]);
+
+  const handleDismissWithReset = useCallback(() => {
+    resetToOriginalState();
+    handleDismiss();
+  }, [resetToOriginalState, handleDismiss]);
+
+  const handleSave = useCallback(() => {
```

Comment:
Unfortunately we cannot just update everything only on save button click, we wont be able to support dynamic content with this model...
the parameter changes need to be present in the store to use the dynamic pipeline which gets the dependent parameters from store to use the commited value... this would mean we have to redo dynamic pipeline which makes this WI very complex.

Just for save changes we will have to keep the original state of parameters in components and if user doesnt save we override the parameter values in store using that original state. if they do save changes we can update conditional visibility but values need to be updated as user is updating them.